### PR TITLE
Ensure Byte use is consistent with DAP2 standards (solves #5)

### LINKED ISSUE #5: Check byte use
I noticed in some places of the code byte is signed, while in others it's unsigned. I should verify how the DAP spec defines it (unsigned, I think) and ensure the proper type in the code.

### PR DESCRIPTION
**Pull request by laliberte**
This PR follows #91 and aims to close #5. It provides a much more systematic way of encoding data to DODS and decoding DODS data to Numpy. Along the way, it clarifies the Byte usage. In DAP2, there is no Byte type, only Unsigned Byte type. This means that to encode Signed Bytes to DODS, they have to be upconverted to Int16 and then encoded to DODS. With this PR this is now done transparently.

### CODE REVIEW

**Comment by laliberte** [comment]
@jameshiebert could you review this PR? It is kind of a big one and it should bring stability to the code, enabling some more streamlining in the future. In particular, this one is kind of useful in one is to tackle #76.

**Comment by jameshiebert** [comment]
Sure, I can have a look at this tomorrow. Need a bit of time to concentrate on it... does look like a big one!

**Comment by laliberte** [comment]
Ok, I'll try to be available tomorrow so that I can answer some of your questions as they come.

File: src/pydap/responses/dods.py [hunk 1]
```diff
@@ -53,12 +64,16 @@ def __init__(self, dataset):
             self.headers.append(('Content-length', str(length)))
 
     def __iter__(self):
+        length = 0
```
Comment (jameshiebert) [inline src/pydap/responses/dods.py]: `length` doesn't ever get used. Do we need this?

File: src/pydap/responses/dods.py [hunk 2]
```diff
@@ -108,12 +126,21 @@ def _sequencetype(var):
...
-                dtype = ','.join(types).format(*padded)
-
-            if dtype not in cache:
-                cache[dtype] = np.zeros((1,), dtype=dtype)
-            cache[dtype][:] = tuple(record)
-            yield cache[dtype].tostring()
+                DAP2_dtype_str = ','.join(DAP2_types).format(*padded)
+
+            if DAP2_dtype_str not in cache:
+                # Remember that DAP2_dtype is a (possibly composite)
```
Comment (jameshiebert) [inline src/pydap/responses/dods.py]: Thanks for commenting and clarifying the intent of this section of code.

File: src/pydap/parsers/dds.py [hunk 3]
```diff
@@ -8,25 +8,22 @@
...
-    'int16':   np.dtype(">i4"),
-    'uint16':  np.dtype(">u4"),
-    'int32':   np.dtype(">i4"),
-    'uint32':  np.dtype(">u4"),
-    'float32': np.dtype(">f4"),
-    'float64': np.dtype(">f8"),
-    'string':  np.dtype(STRING),  # default is "|S128"
-    'url':     np.dtype(STRING),
-    }
+from ..lib import (quote, LOWER_DAP2_TO_NUMPY_PARSER_TYPEMAP)
```
Comment (jameshiebert) [inline src/pydap/parsers/dds.py]: Check your line endings in this file. Looks like you have \r\n (that should just \n). Just run this through dos2unix or something.

File: src/pydap/responses/dods.py (hunk 2 above)
Comment (laliberte) [inline src/pydap/responses/dods.py]: This was sufficiently confusing and I thought anyone else would be as confused. At least now I've made sure it is pretty clear...

**Comment by laliberte** [comment]
@jameshiebert this PR seems fairly completed to me. Should I merge?

**Comment by laliberte** [comment]
I've done one last thing: reintroduce client tests that were previously skipped that were related to datatypes.

**Comment by jameshiebert** [comment]
Yep, this looks great. Thanks!

**Pull request merged**